app/controllers/items_controller.rb

```diff
@@ -2,18 +2,86 @@ class ItemsController < OauthController
   before_action :login
 
   def refile
+    @refile_request = RefileRequest.new
+    # TODO: Finish this in subsequent branch.
+    if params[:page] && params[:per_page]
+      @offset = ( params[:page].to_i - 1 ) * params[:per_page].to_i
+    else
+      @offset = 0
+    end
+
+    date_start = params[:date_start].present? ? params[:date_start] : nil
+    date_end = params[:date_end].present? ? params[:date_end] : nil
+    @refiles = get_refiles(date_start,date_end,params[:page],params[:per_page])
   end
   
   def update_metadata
+    @barcodes = params[:notice].present? && params[:notice]["barcodes"].present? ? params[:notice]["barcodes"] : []
+    @protect_cgd = params[:notice].present? && params[:notice]["protect_cgd"].present? ? params[:notice]["protect_cgd"] : false
   end
 
   def send_metadata
-    barcodes = params[:barcodes]
-    puts barcodes
-    flash[:notice] = "Metadata updated for #{barcodes}"
-    redirect_to request.referrer
+    # don't let it fool you. params[:barcodes] is text.
+    barcodes = params[:barcodes].strip.split(/[\W\s]+/)
+    message = Message.new(barcodes: barcodes, protect_cgd: params[:protect_cgd], action: 'update')
+    if message.valid?
+      message.send_update_message_to_sqs
+      invalid_barcodes = []
+    elsif message.valid_barcodes.present? 
+      invalid_barcodes = message.invalid_barcodes
+      message.barcodes = message.valid_barcodes
+      message.send_update_message_to_sqs
+    else
+      invalid_barcodes = barcodes
+    end
+    
+    protect_cgd = message.errors.present? ? params[:protect_cgd] : false
+    
+    if barcodes.blank?
+      flash[:error] = "Please enter one or more barcodes."
+    elsif invalid_barcodes.present? && message.valid_barcodes.blank?
+      flash[:error] = "The barcode(s) remaining are invalid; each barcode must be 14 numerical digits in length. Separate barcodes using commas or returns (new lines)."
+    elsif invalid_barcodes.present? && message.valid_barcodes.present?
+      flash[:notice] = "Some barcodes were submitted."
+      flash[:error] = "The barcode(s) remaining are invalid; each barcode must be 14 numerical digits in length. Separate barcodes using commas or returns (new lines)."
+    end
+    if message.valid_barcodes.present? && invalid_barcodes.blank?
+      flash[:success] = "The barcode(s) have been submitted for processing."
+    end
+    redirect_to action: 'update_metadata', notice: { barcodes: invalid_barcodes, protect_cgd: protect_cgd}
```

Comment:
I think we can find a way to smuggle these values back to the "edit" form without using the hash.
Let's look into a way of using an `ActiveModel` object and either `form_for` or the `value: ` option in the form tags to have the form get its values directly from our instance of `message.`

Reply:
Ahhh - this isn't actually using the flash - it's turning these into parameters. 
For example, after a bad submit you are redirected to:

```
http://local.nypl.org:3001/update_metadata?notice[barcodes][]=sdahusdhiu¬ice[protect_cgd]=
```